src/mpi/request/mpir_request.c

```diff
@@ -72,8 +75,20 @@ int MPIR_Request_completion_processing(MPIR_Request * request_ptr, MPI_Status *
                 MPII_SENDQ_FORGET(request_ptr);
                 break;
             }
-        case MPIR_REQUEST_KIND__RECV:
         case MPIR_REQUEST_KIND__COLL:
+            {
+                MPII_Coll_req_t *coll = &request_ptr->u.nbc.coll;
+                if (coll->host_recvbuf)
+                    MPIR_Localcopy(coll->host_recvbuf, coll->count, coll->datatype,
+                                   coll->user_recvbuf, coll->count, coll->datatype);
+                MPIR_Coll_host_buffer_free(coll->host_sendbuf, coll->host_recvbuf);
+                MPIR_Datatype_release_if_not_builtin(coll->datatype);
+
+                MPIR_Request_extract_status(request_ptr, status);
+                mpi_errno = request_ptr->status.MPI_ERROR;
+                break;
+            }
```

Comment:
`MPIR_Request_completion_processing` is not protected by VCI lock, mostly because the other processing doesn't really modify the request objects. I think you need add the same request VCI lock here.

Reply:
Is request completion processing now executed under a lock by the progress engine? If this function is not executed under a lock, adding one here won't fix the race condition I am concerned about.

Reply:
If that is the case, then we need to pre-create the request and stash these fields in it before issuing the collective. Need to look at the algorithms to see how much churn that will add.

Reply:
Request completion is done by MPI function such as `MPI_Test`, `MPI_Wait`. I guess because you haven't returned the request to the user yet, none of the wait/test function will be completing this request, which means maybe you don't need critical section at all. Let's double check...

Reply:
... yeah, I think you are safe without any critical sections.

Reply:
I guess it comes down to: if a collective schedule is started, will it be progressed by MPICH without further user intervention? Will dig in some more...

Reply:
Hmm, yes. Looks like you are right. Completion processing only occurs if the user tests or waits on the request. That should simplify these changes a bit. New version coming shortly...